File: python/complex-numbers/complex_numbers.py
import math


class ComplexNumber:
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        real = self.real + other.real
        imaginary = self.imaginary + other.imaginary

        return ComplexNumber(real, imaginary)

    def __mul__(self, other):
        real = self.real * other.real - self.imaginary * other.imaginary
        imaginary = self.imaginary * other.real + self.real * other.imaginary

        return ComplexNumber(real, imaginary)

    def __sub__(self, other):
        real = self.real - other.real
        imaginary = self.imaginary - other.imaginary

        return ComplexNumber(real, imaginary)

    def __truediv__(self, other):
        real = (self.real * other.real + self.imaginary *
                other.imaginary) / (abs(other) * abs(other))
        imaginary = (self.imaginary * other.real - self.real *
                     other.imaginary) / (abs(other) * abs(other))

        return ComplexNumber(real, imaginary)

    def __abs__(self):
        square_sum = self.real * self.real + self.imaginary * self.imaginary

        return math.sqrt(square_sum)

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        real = math.e ** self.real * math.cos(self.imaginary)
        imaginary = math.e ** self.real * math.sin(self.imaginary)

        return ComplexNumber(real, imaginary)

File: haskell/complex-numbers/src/ComplexNumbers.hs
module ComplexNumbers
  ( Complex
  , conjugate
  , abs
  , exp
  , real
  , imaginary
  , mul
  , add
  , sub
  , div
  , complex
  ) where

import           Prelude                 hiding ( abs
                                                , div
                                                , exp
                                                )
import qualified Prelude                       as P

-- Data definition -------------------------------------------------------------
data Complex a = Complex a a
  deriving (Eq, Show)

complex :: (a, a) -> Complex a
complex = uncurry Complex

-- unary operators -------------------------------------------------------------
conjugate :: Num a => Complex a -> Complex a
conjugate (Complex a b) = Complex a (-b)

abs :: Floating a => Complex a -> a
abs (Complex a b) = sqrt (a * a + b * b)

real :: Num a => Complex a -> a
real (Complex a _) = a

imaginary :: Num a => Complex a -> a
imaginary (Complex _ b) = b

exp :: Floating a => Complex a -> Complex a
exp (Complex a b) = Complex (P.exp a * cos b) (P.exp a * sin b)

-- binary operators ------------------------------------------------------------
mul :: Num a => Complex a -> Complex a -> Complex a
mul (Complex a b) (Complex c d) = Complex (a * c - b * d) (b * c + a * d)

add :: Num a => Complex a -> Complex a -> Complex a
add (Complex a b) (Complex c d) = Complex (a + c) (b + d)

sub :: Num a => Complex a -> Complex a -> Complex a
sub (Complex a b) (Complex c d) = Complex (a - c) (b - d)

div :: Fractional a => Complex a -> Complex a -> Complex a
div (Complex a b) (Complex c d) = Complex ((a * c + b * d) / (c * c + d * d))
                                          ((b * c - a * d) / (c * c + d * d))
